### PR TITLE
feat(Core/Player): implement Spell Queue

### PR DESCRIPTION
**Pull request by sogladev**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

Implement a system to queue and process spell/item cast requests.
- Queue can store multiple casts prioritized by time requested (FIFO), allowing interactions such as using an item + casting a spell immediately after a cast finishes
- Window is set at [400ms, based on this discussion](https://github.com/azerothcore/azerothcore-wotlk/issues/9300#issuecomment-977211452)

Uses [similar flow logic as TC master](https://github.com/TrinityCore/TrinityCore/pull/29409) and [Cata preservation](https://github.com/The-Cataclysm-Preservation-Project/TrinityCore/commit/4430863a1e9a787a3a702b58adaf4c7bb639fa29#diff-b4a23722c5b449d326b1c0e5a75f3b51de931e42e688793fd9f1d233536a02c6R17)

One key difference is that the 335a client does not send `CMSG_CAST_SPELL`  packets while on GCD, making queueing spells during casts that last the duration of GCD not possible, such as instant cast corruption. Only spells queued during the end of a current spell cast (<400ms) will be queued. 

For item casts, `CMSG_USE_ITEM` are sent when on GCD, so items can be queued during GCD e.g., casting a 1.5s spell and queuing a potion is possible

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9300

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [x] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

tested rotations of some specs:
- affli warlock
- moonkin
- surv hunter
- warrior

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PRs page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by walkline** [comment]
As I understood it works in the next way:
IMG_9242
Every player has its own independent queue with its own time frame for processing.

I assume it should be global for all players, like this:
IMG_9243

My assumption is incorrect?

**Comment by Tereneckla** [comment]
No, you are mistaken. The first graphic is correct

**Comment by Tereneckla** [comment]
No, it's just a way to circumvent input lag when casting spells. 400ms window to enqueue the spell so that it starts casting immediately after your last

**Comment by sudlud** [comment]
Can you merge master please

File: src/server/game/Handlers/SpellHandler.cpp [hunk 1]
```diff
@@ -85,6 +85,41 @@ void WorldSession::HandleUseItemOpcode(WorldPacket& recvPacket)
...
+
+    if (!spellInfo)
+    {
+        LOG_ERROR("network.opcode", "WORLD: unknown spell id {}", spellId);
+        recvPacket.rfinish(); // prevent spam at ignore packet
+        return;
+    }
+
+    // fail if we are cancelling pending request
+    if (_player->SpellQueue.size())
```
Comment (Winfidonarleyan) [inline src/server/game/Handlers/SpellHandler.cpp:98]: Need check via empty

**Comment by Grimdhex** [comment]
Great job on the feature. I've played with this PR for 15min and I didn't see any particular issue. Work as intended with spells and items. 

Just why disable it by default? To me, as it's a blizz feature, it should be enabled (and let the players disable it if they don't want it).

**Comment by heyitsbench** [comment]
I do appreciate the feature being put behind a config option, but I do agree with defaulting it to enabled, as it is a Blizzlike feature.

File: src/server/apps/worldserver/worldserver.conf.dist [hunk 2]
```diff
@@ -2163,6 +2163,21 @@ NoResetTalentsCost = 0
 
 ToggleXP.Cost = 100000
 
+#
+#    SpellQueue.Enabled
+#        Description: Enable SpellQueue.
+#        Default:     0 - (Disabled)
+#                     1 - (Enabled)
+
+SpellQueue.Enabled = 0
```
Comment (Nyeriah) [inline src/server/apps/worldserver/worldserver.conf.dist:2172]:
```suggestion
SpellQueue.Enabled = 1
```

File: src/server/game/World/World.cpp [hunk 3]
```diff
@@ -1487,6 +1487,10 @@ void World::LoadConfigSettings(bool reload)
     // Realm Availability
     _bool_configs[CONFIG_REALM_LOGIN_ENABLED] = sConfigMgr->GetOption<bool>("World.RealmAvailability", true);
 
+    // SpellQueue
+    _bool_configs[CONFIG_SPELL_QUEUE_ENABLED] = sConfigMgr->GetOption<bool>("SpellQueue.Enabled", false);
```
Comment (Nyeriah) [inline src/server/game/World/World.cpp:1491]:
```suggestion
    _bool_configs[CONFIG_SPELL_QUEUE_ENABLED] = sConfigMgr->GetOption<bool>("SpellQueue.Enabled", true);
```

**Review by Nyeriah — APPROVED** [review]

**Review by Grimdhex — APPROVED** [review]

**Comment by boom8866** [comment]
It has a problem that when the player casts a spell too quickly, the second spell is not cast in the normal sequence, and keeping cycling through the first spell or other spells will result in the spell not being released, or even if the player presses the spell faster, it will result in the loss of the spell

**Comment by kissingers** [comment]
> It has a problem that when the player casts a spell too quickly, the second spell is not cast in the normal sequence, and keeping cycling through the first spell or other spells will result in the spell not being released, or even if the player presses the spell faster, it will result in the loss of the spell

Yes, this is the big problem, many players say they can not cast spell right as they think, some spell always can not cast if use default as 1

**Comment by Nyeriah** [comment]
It defaults to ENABLED because this is not a "custom" feature but actually part of the core. Users can either disable it if they want to or use a lower ms delay.

**Comment by kissingers** [comment]
> It defaults to ENABLED because this is not a "custom" feature but actually part of the core 335 features. Users can either disable it if they want to or use a lower ms delay.


I know that enabling this option is quite good and can increase the player's DPS, but there's a problem: if you keep pressing multiple skills, you can't cast the desired skill when needed. For example, with a warlock's abilities, if you keep pressing Incinerate, you won't be able to cast Chaos Bolt when you want to. so if this problem can fix, it will more good.

**Comment by Grimdhex** [comment]
> > It defaults to ENABLED because this is not a "custom" feature but actually part of the core 335 features. Users can either disable it if they want to or use a lower ms delay.
> 
> 
> 
> 
> 
> I know that enabling this option is quite good and it can increase the player's DPS. However, there's a problem: if you keep pressing multiple skills, you can't cast the desired skill when needed. For example, with a warlock's abilities, if you keep pressing Incinerate, when the skill Chaos Bolt becomes available, pressing it won't cast it; instead, it will continue to cast Incinerate.
> 
> 
> 
> It would be perfect if the system could be improved so that if no new skill is pressed, it continues to queue other skills. But if a new skill is pressed and the server receives it in time, it would be inserted at the front of the queue for casting.
> 
> 
> 
> just same as  [boom8866] description before.

Probably a bug so please open an issue on it.

**Comment by Tereneckla** [comment]
You can't queue up spells for multiple casts so I don't see how spamming a spell would hinder you from casting another spell.
Unless your players are spamming Incinerate in the 400ms window only to then realize Chaos Bolt goes of CD. I'd quite frankly call that a skill issue.

**Pull request merged**